File: internal/detect/detect.go
// Package detect finds secrets in request bodies. Detection rules are
// supplied by Providers (see provider.go); a Shannon-entropy heuristic
// catches unknown high-entropy tokens when enabled in Config.
package detect

import (
	"bytes"
	"cmp"
	"fmt"
	"regexp"
	"slices"
	"strings"

	"github.com/sourcegraph/conc/pool"
)

// parallelScanMinBody is the body size (bytes) at or above which Scan
// fans the rule set out across goroutines. Below it the goroutine
// scheduling cost outweighs the regex work, so Scan stays on the
// sequential path.
const parallelScanMinBody = 4096

// Finding is one detected secret.
type Finding struct {
	Value    string
	Rule     string
	Severity string
}

// Config controls the optional entropy layer.
type Config struct {
	Entropy          bool
	EntropyThreshold float64 // bits per character; 0 applies the default
	EntropyMinLen    int     // 0 applies the default
}

// Detector scans bytes for secrets.
type Detector struct {
	rules []compiledRule
	cfg   Config
}

type compiledRule struct {
	name        string
	severity    string
	group       int
	re          *regexp.Regexp
	prefix      string
	prefixBytes []byte // prefix as []byte, for the Scan literal-prefix gate
}

// New builds a Detector. When called with no providers, it falls back
// to DefaultProviders(). Providers are flattened in declaration order;
// when two rules match the same byte span, the first one wins.
func New(cfg Config, providers ...Provider) (*Detector, error) {
	if cfg.EntropyThreshold == 0 {
		cfg.EntropyThreshold = 4.0
	}
	if cfg.EntropyMinLen == 0 {
		cfg.EntropyMinLen = 24
	}
	if len(providers) == 0 {
		providers = DefaultProviders()
	}

	var total int
	for _, p := range providers {
		total += len(p.Rules())
	}
	compiled := make([]compiledRule, 0, total)
	for _, p := range providers {
		for _, r := range p.Rules() {
			if r.Regex == nil {
				return nil, fmt.Errorf("detect: provider %q: rule %q has nil regex", p.Name(), r.Name)
			}
			prefix, _ := r.Regex.LiteralPrefix()
			compiled = append(compiled, compiledRule{
				name:        r.Name,
				severity:    r.Severity,
				group:       r.Group,
				re:          r.Regex,
				prefix:      prefix,
				prefixBytes: []byte(prefix),
			})
		}
	}
	return &Detector{rules: compiled, cfg: cfg}, nil
}

// RuleCount returns the number of compiled detection rules across all
// providers.
func (d *Detector) RuleCount() int { return len(d.rules) }

// Scan returns every distinct secret found in body, sorted by value.
// Regex rules always run; the entropy heuristic runs only when enabled
// in Config.
func (d *Detector) Scan(body []byte) []Finding {
	seen := make(map[string]Finding, len(d.rules))

	applyRule := func(r compiledRule) []Finding {
		if len(r.prefixBytes) > 0 && !bytes.Contains(body, r.prefixBytes) {
			return nil
		}
		var local []Finding
		for _, m := range r.re.FindAllSubmatch(body, -1) {
			if r.group >= len(m) {
				continue
			}
			v := string(m[r.group])
			if v == "" {
				continue
			}
			local = append(local, Finding{Value: v, Rule: r.name, Severity: r.severity})
		}
		return local
	}

	merge := func(findings []Finding) {
		for _, f := range findings {
			if _, ok := seen[f.Value]; !ok {
				seen[f.Value] = f
			}
		}
	}

	if len(body) >= parallelScanMinBody {
		p := pool.NewWithResults[[]Finding]()
		for _, r := range d.rules {
			p.Go(func() []Finding { return applyRule(r) })
		}
		for _, findings := range p.Wait() {
			merge(findings)
		}
	} else {
		for _, r := range d.rules {
			merge(applyRule(r))
		}
	}

	if d.cfg.Entropy {
		for _, tok := range entropyTokens(body, d.cfg.EntropyThreshold, d.cfg.EntropyMinLen) {
			if _, ok := seen[tok]; !ok {
				seen[tok] = Finding{Value: tok, Rule: "entropy", Severity: "medium"}
			}
		}
	}
	out := make([]Finding, 0, len(seen))
	for _, f := range seen {
		out = append(out, f)
	}
	slices.SortFunc(out, func(a, b Finding) int { return cmp.Compare(a.Value, b.Value) })
	return out
}

// LiteralPrefixLen returns the byte length of the longest rule literal
// prefix that value begins with — used to keep a credential's
// recognizable prefix intact when garbling. Returns 0 when no rule
// prefix matches.
func (d *Detector) LiteralPrefixLen(value string) int {
	best := 0
	for _, r := range d.rules {
		if len(r.prefix) > best && strings.HasPrefix(value, r.prefix) {
			best = len(r.prefix)
		}
	}
	return best
}

File: hooks/detect.ts
// Detection: the regex rules always run. The Shannon-entropy heuristic runs
// only when the `entropy` plugin option turns it on, because on ordinary
// prompt text it reports hashes, base64 blocks and git SHAs as secrets.

import { RULES } from './rules'

export type DetectConfig = {
  entropy: boolean
  entropyThreshold: number // bits per character
  entropyMinLen: number
}

export const DEFAULT_DETECT: DetectConfig = {
  entropy: false,
  entropyThreshold: 4.0,
  entropyMinLen: 24,
}

const QUANTIFIER = new Set(['?', '*', '+', '{'])
const META = new Set(['[', '(', ')', '{', '}', '?', '*', '+', '|', '^', '$', '.'])

/**
 * The fixed text a pattern must begin with, the JS answer to Go's
 * `Regexp.LiteralPrefix`. Used for two things: a cheap `includes` gate before
 * running a rule, and the number of leading bytes a garble keeps verbatim so a
 * fake still reads as that vendor's credential.
 */
export function literalPrefix(source: string): string {
  let out = ''
  for (let i = 0; i < source.length; i++) {
    const c = source[i]!
    if (c === '\\') {
      const escaped = source[i + 1]
      // \d \w \s \b … are classes, not literals.
      if (escaped === undefined || /[a-zA-Z0-9]/.test(escaped)) break
      if (QUANTIFIER.has(source[i + 2] ?? '')) break
      out += escaped
      i++
      continue
    }
    if (META.has(c)) break
    // A quantifier makes the character it follows optional or repeated, so it
    // is not part of a fixed prefix.
    if (QUANTIFIER.has(source[i + 1] ?? '')) break
    out += c
  }
  return out
}

const PREFIXES: readonly string[] = RULES.map((r) => literalPrefix(r.re.source))

export type Finding = {
  value: string
  rule: string
  severity: string
}

/**
 * Every distinct secret in `text`. First rule to claim a value wins, matching
 * first-rule-wins dedup.
 */
export function scan(text: string, cfg: DetectConfig = DEFAULT_DETECT): Finding[] {
  const seen = new Map<string, Finding>()

  RULES.forEach((r, i) => {
    const prefix = PREFIXES[i]!
    if (prefix.length > 0 && !text.includes(prefix)) return
    r.re.lastIndex = 0
    for (const m of text.matchAll(r.re)) {
      const value = m[r.group]
      if (!value) continue
      if (!seen.has(value)) {
        seen.set(value, { value, rule: r.name, severity: r.severity })
      }
    }
  })

  if (cfg.entropy) {
    for (const token of entropyTokens(text, cfg.entropyThreshold, cfg.entropyMinLen)) {
      if (!seen.has(token)) {
        seen.set(token, { value: token, rule: 'entropy', severity: 'medium' })
      }
    }
  }

  return [...seen.values()].sort((a, b) => (a.value < b.value ? -1 : a.value > b.value ? 1 : 0))
}

/**
 * The length of the longest rule prefix `value` begins with; 0 when none
 * matches.
 */
export function literalPrefixLen(value: string): number {
  let best = 0
  for (const p of PREFIXES) {
    if (p.length > best && value.startsWith(p)) best = p.length
  }
  return best
}

// ------------------------------------------------------------------ entropy

function shannon(s: string): number {
  if (s.length === 0) return 0
  const freq = new Map<string, number>()
  for (const ch of s) freq.set(ch, (freq.get(ch) ?? 0) + 1)
  let h = 0
  for (const count of freq.values()) {
    const p = count / s.length
    h -= p * Math.log2(p)
  }
  return h
}

const TOKEN = /[0-9a-zA-Z\-_.+/=]{1,}/g

function entropyTokens(text: string, threshold: number, minLen: number): string[] {
  const out: string[] = []
  TOKEN.lastIndex = 0
  for (const m of text.matchAll(TOKEN)) {
    const token = m[0]
    if (token.length >= minLen && shannon(token) >= threshold) out.push(token)
  }
  return out
}
